doc/source/architecture.rst

```diff
@@ -0,0 +1,213 @@
+Verbs include GET, PUT, UPDATE, DELETE, and POST, while resources are any
+uniquely identifiable entity.
+
+REST can scale because servers can make assumptions about the data it serves
+based on the verb and resource. For instance, if a GET was done on the resource
+"/a/b/c", and then no mutating verbs (PUT, UPDATE, DELETE, POST) have been run
+on that or any child resource (such as "/a/b/c/d"), the server can reuse the
+result of the previous request rather than recompute the result. When using the
+HTTP protocol, this is called HTTP caching and is a fundamental property of the
+internet.
+
+Remote method invocation
+------------------------
+
+In the 1990s, OOP exploded in popularity. With it, came a very tempting idea.
+OOP allows programmers to provide abstractions on top of data in their code.
+It was thought that even the difference between RPC calls and calls made in
+a program's address space could be abstracted. The user of the library does
+not need to know whether an object exists remotely or locally, the call can
+just be made in the same way, and to scale the system to be distributed over a
+network, some middleware would be responsible for load-balancing and allocating
+these objects remotely.
+
+This technique is known as Remote Method Invocation (RMI). RMI was widely
+implemented, using CORBA, DCOM, Remoting (.NET), and Java RMI. However, this
+approach has fallen out of favor with the rise of the internet, as it was
+observed that this paradigm does not scale. As a practical example, recent
+versions of .NET do not implement the Remoting library, and COM/DCOM are not
+used in modern web stacks.
+
+An illuminating discussion of the problems with RMI can be found at
+https://martinfowler.com/articles/distributed-objects-microservices.html
+
+Remote mechanical
+=================
+
+Mechanical's official API is that of an object model, and PyMechanical provides
+exactly that API to Python. Because Object Models are not suitable as remote
+APIs, PyMechanical does not provide that API in a remote fashion. This is why
+the remote session API is based on strings, while the embedded instance API can
+provide the Mechanical API directly to Python.
+
+Another remote API
+------------------
+
+Another remote API for Mechanical is definitely possible to provide, so long as
+it is based on message passing or REST. In fact, Mechanical uses a REST API
+internally as part of its GUI. This is not the official Mechanical API and is
+not currently documented. But such an API is not an immediate goal of the
+PyMechanical project.
+
+Using PyMechanical in a distributed system
+------------------------------------------
+It is still possible to build a distributed system where Mechanical is run
+remotely using the embedded instance of Mechanical in PyMechanical. For this to
+work, Python itself would run remotely, and therefore the embedded instance
+would run remotely. The communication across the network in that system would
+be done in Python.
+
```

Comment:
You can build a distributed system to run the application remotely by running Python remotely when it includes an embedded instance of the application. All network communication is performed in Python.